### PR TITLE
Fix regression where '?' is not removed from boolean syntactic sugar on attributes.

### PR DESCRIPTION
**Pull request by NullVoxPopuli**
#### Purpose

Pre 0.10, `?` was removed from attributes.

#### Changes

Added Test

#### Caveats

What about methods with `?` -- the ones that aren't just syntactic sugar?
This may result in checking if the object responds_to a method without the question mark -- which may lead to unexpected behavior.

#### Related GitHub issues

https://github.com/rails-api/active_model_serializers/issues/605

### CODE REVIEW

**Comment by NullVoxPopuli** [comment]
I guess the ? could just be removed in the rendered output, and if there are collisions, write to the log?

like if someone did 

```ruby
attribute :boolean, :boolean?
```

File: test/serializers/attribute_test.rb [hunk 1]
```diff
@@ -81,6 +81,23 @@ def id
         assert_equal('custom', hash[:blog][:id])
       end
 
+      def test_id_with_question_mark_has_question_mark_removed
+        serializer = Class.new(ActiveModel::Serializer) do
+          # TODO: if we were to remove the question mark automatically,
+          #       what do we do if we want to call a method that has a
+          #       question mark.
+          #       e.g.:
+          #               def my_question?; ....; end
```
Comment (bf4) [inline test/serializers/attribute_test.rb]: In rails boolean fields work w and w out the ?. String inquirer takes care of it

Comment (bf4) [inline test/serializers/attribute_test.rb]: Best practice is just don't use a ? In your serializer

? Is def verboten in json api

Comment (NullVoxPopuli) [inline test/serializers/attribute_test.rb]: yeah, but I mean if someone wanted to use a method in their model rather than just a db-backed field.

e.g.:
```ruby
2.2.3 :001 > def test?; 'hi'; end
 => :test? 
2.2.3 :002 > test
ArgumentError: wrong number of arguments (0 for 2..3)
	from (irb):2:in `test'
	from (irb):2
	from /home/preston/.rvm/rubies/ruby-2.2.3/bin/irb:11:in `<main>'
2.2.3 :003 > test?
 => "hi" 
2.2.3 :004 > 
```

Comment (bf4) [inline test/serializers/attribute_test.rb]: > like if someone did
> attribute :boolean, :boolean?

Don't do that.  When would that ever make sense?

Comment (bf4) [inline test/serializers/attribute_test.rb]: If your ? Method doesn't return a boolean that's a bug waiting to happen

Comment (NullVoxPopuli) [inline test/serializers/attribute_test.rb]: well, if that's the case, then I guess we should remove it, and just note in our docs, about the possibility of a collision, if someone were silly enough to try rendering both of `:boolean, :boolean?` where both of those point to different data.

maybe it's just an edge case, and it won't be that big of a deal.

Comment (bf4) [inline test/serializers/attribute_test.rb]: I would say they are asking for trouble and should get it.  They can always set a key:

Comment (NullVoxPopuli) [inline test/serializers/attribute_test.rb]: good point.

Ok, I'll implement the fix for this then.

Comment (bf4) [inline test/serializers/attribute_test.rb]: Still is a good test to have to demonstrate behavior.  Need to confirm json api adapter strips it

Comment (NullVoxPopuli) [inline test/serializers/attribute_test.rb]: just json api adapter? or all adapters?

**Comment by NullVoxPopuli** [comment]
Fixed

**Comment by NullVoxPopuli** [comment]
squashed

**Comment by remear** [comment]
Looks good to me

File: lib/active_model/serializer/attributes.rb [hunk 2]
```diff
@@ -17,6 +17,8 @@ module Attributes
         def attributes(requested_attrs = nil, reload = false)
           @attributes = nil if reload
           @attributes ||= self.class._attributes_data.each_with_object({}) do |(key, attr), hash|
+            key = key.to_s.gsub(/\?\z/, '').to_sym
```
Comment (groyoh) [inline lib/active_model/serializer/attributes.rb]: Would it be possible to move that at https://github.com/rails-api/active_model_serializers/blob/master/lib/active_model/serializer/attributes.rb#L57? That way it's done at the serializer definition and not everytime we call the `#attributes` method.

Comment (NullVoxPopuli) [inline lib/active_model/serializer/attributes.rb]: yes! good catch. Pushed :-)

File: lib/active_model/serializer/attributes.rb [hunk 3]
```diff
@@ -55,6 +55,9 @@ def attributes(*attrs)
         #     end
         def attribute(attr, options = {}, &block)
           key = options.fetch(:key, attr)
+          # replace ? at the end of an attribute, as JSON API disallows the use
+          # of ? at the end of keys.
+          key = key.to_s.gsub(/\?\z/, '').to_sym
```
Comment (groyoh) [inline lib/active_model/serializer/attributes.rb]: I'll be picky here but I usually tend to avoid using `gsub` when `sub` can be used: ` key = key.to_s.sub(/\?\z/, '').to_sym` :grin:

**Comment by groyoh** [comment]
Looks good :+1: feel free to listen to my picky comment :wink:

File: lib/active_model/serializer/attributes.rb (hunk 3 above)
Comment (NullVoxPopuli) [inline lib/active_model/serializer/attributes.rb]: I didn't know sub could handle regex -- that's why I use it. 

http://ruby-doc.org/core-2.2.0/String.html#method-i-sub
well, whaddayaknow.

oh, it looks like gsub replaces all occurrences, and sub replaces just the first occurrence.

that leads me to believe sub would be slightly faster. I'll swap it out.

**Comment by NullVoxPopuli** [comment]
@groyoh just swapped out gsub for sub. thanks for the tip!

**Comment by bf4** [comment]
'tr' is actually faster for single char and '?' Is only legal once and at the end

B mobile phone

File: lib/active_model/serializer/attributes.rb (hunk 3 above)
Comment (bf4) [inline lib/active_model/serializer/attributes.rb]: Gem install fasterer

B mobile phone

**Comment by NullVoxPopuli** [comment]
@bf4, actually, it looks like sub is faster.

I made a benchmark
```ruby
require 'benchmark'

attribute = 'some_boolean?'
n = 1000000
Benchmark.bm do |benchmark|
  benchmark.report { n.times{ attribute.sub(/\?\z/, '') }}
  benchmark.report { n.times{ attribute.gsub(/\?\z/, '') }}
  benchmark.report { n.times{ attribute.tr('?', '') }}
end
```
results:
```
⬡ v4.3.1 💎  2.2.3 $ ruby test.rb
       user     system      total        real
   0.530000   0.000000   0.530000 (  0.550814)
   1.450000   0.010000   1.460000 (  1.472772)
   0.550000   0.000000   0.550000 (  0.557974)

```

**Comment by NullVoxPopuli** [comment]
just did another one with the mutating versions of each of those:

```ruby
require 'benchmark'

attribute = 'some_boolean?'
n = 1000000
Benchmark.bm do |x|
  x.report('sub  ') { n.times{ attribute.sub(/\?\z/, '') }}
  x.report('gsub ') { n.times{ attribute.gsub(/\?\z/, '') }}
  x.report('tr   ') { n.times{ attribute.tr('?', '') }}
  x.report('sub! ') { n.times{ attribute.sub!(/\?\z/, '') }}
  x.report('gsub!') { n.times{ attribute.gsub!(/\?\z/, '') }}
  x.report('tr!  ') { n.times{ attribute.tr!('?', '') }}
end

```


results:
```
⬡ v4.3.1 💎  2.2.3 $ ruby test.rb
       user     system      total        real
sub    0.530000   0.000000   0.530000 (  0.545544)
gsub   1.430000   0.000000   1.430000 (  1.478049)
tr     0.540000   0.000000   0.540000 (  0.563370)
sub!   0.350000   0.040000   0.390000 (  0.390127)
gsub!  0.430000   0.000000   0.430000 (  0.431214)
tr!    0.480000   0.000000   0.480000 (  0.486672)

```

**Comment by bf4** [comment]
Try w benchmark-ips

**Comment by NullVoxPopuli** [comment]
ok, new results:

```
⬡ v4.3.1 💎  2.2.3 $ ruby test.rb
Calculating -------------------------------------
               sub      37.765k i/100ms
               gsub     24.482k i/100ms
               tr       40.036k i/100ms
               sub!     42.971k i/100ms
               gsub!    41.631k i/100ms
               tr!      42.918k i/100ms
-------------------------------------------------
               sub        1.832M (± 8.4%) i/s -      9.064M
               gsub       1.858M (± 8.9%) i/s -      9.156M
               tr         1.589M (± 6.1%) i/s -      7.927M
               sub!       2.122M (± 6.6%) i/s -     10.528M
               gsub!      2.071M (± 8.7%) i/s -     10.241M
               tr!        1.756M (± 7.9%) i/s -      8.712M

Comparison:
               sub! :  2122277.9 i/s
               gsub!:  2070864.3 i/s - 1.02x slower
               gsub :  1857651.1 i/s - 1.14x slower
               sub  :  1832473.0 i/s - 1.16x slower
               tr!  :  1756340.4 i/s - 1.21x slower
               tr   :  1589408.6 i/s - 1.34x slower

```

via

```ruby
require 'benchmark/ips'

attribute = 'some_boolean?'

Benchmark.ips do |x|
  # 5 seconds, warm up for 2 seconds
  x.config(time: 5, warmup: 2)

  x.report('sub  ') { attribute.sub(/\?\z/, '') }
  x.report('gsub ') { attribute.gsub(/\?\z/, '') }
  x.report('tr   ') { attribute.tr('?', '') }
  x.report('sub! ') { attribute.sub!(/\?\z/, '') }
  x.report('gsub!') { attribute.gsub!(/\?\z/, '') }
  x.report('tr!  ') { attribute.tr!('?', '') }

  x.compare!
end
```

File: lib/active_model/serializer/attributes.rb [hunk 4]
```diff
@@ -55,6 +55,9 @@ def attributes(*attrs)
         #     end
         def attribute(attr, options = {}, &block)
           key = options.fetch(:key, attr)
+          # replace ? at the end of an attribute, as JSON API disallows the use
+          # of ? at the end of keys.
+          key = key.to_s.sub(/\?\z/, '').to_sym
```
Comment (bf4) [inline lib/active_model/serializer/attributes.rb]: better if this were in a method someone could monkey patch if they absolutely hate it

better if it only applied in json api

best if it were part of the 'transform keys' logic for json api (which doesn't really exist right now)

Comment (NullVoxPopuli) [inline lib/active_model/serializer/attributes.rb]: > better if this were in a method someone could monkey patch if they absolutely hate it

I agree with it being in its own method (just pushed)

> better if it only applied in json api

do we have a mechanism where this can only apply to JSON API?

> best if it were part of the 'transform keys' logic for json api (which doesn't really exist right now)
 
for later! :-)

**Comment by NullVoxPopuli** [comment]
Closing this in favor of custom key transform.

See: https://github.com/rails-api/active_model_serializers/blob/9e992358d87bb69b52006caf08b6b1093db2b53d/lib/active_model_serializers/key_transform.rb
```ruby
require 'active_support/core_ext/hash/keys'

module ActiveModelSerializers
  module KeyTransform
    module_function

    # Transforms keys to UpperCamelCase or PascalCase.
    #
    # @example:
    #    "some_key" => "SomeKey",
    # @see {https://github.com/rails/rails/blob/master/activesupport/lib/active_support/inflector/methods.rb#L66-L76 ActiveSupport::Inflector.camelize}
    def dashed_without_question(hash)
      hash.deep_transform_keys! { |key| key.to_s.sub(/\?\z/, '').dasherize.to_sym }
    end
    ...
end
```

and then in an initializer:
```ruby
ActiveModelSerializers.config.key_transform = :dashed_without_question
```

/cc @dangalipo, @sija, @pedromtavares from #605

**Comment by dangalipo** [comment]
cool :+1:

**Comment by Sija** [comment]
looks cool with me too! :+1: I wonder tho' is there/would be a possibility of defining `ActiveModelSerializers.config.key_transform` as an array — with transformers executed in a row? I could image it might come helpful at times.

**Comment by NullVoxPopuli** [comment]
@Sija you could just write a custom transform that does your chained operations?

**Comment by Sija** [comment]
@NullVoxPopuli :-1: for modularity and another set of future/possible monkey-patches

**Comment by NullVoxPopuli** [comment]
not all monkey patches are bad. :-\

**Comment by bf4** [comment]
>  you could just write a custom transform that does your chained operations?

@NullVoxPopuli that's a really good idea.  It'd be pretty easy to write and document.

**Comment by Sija** [comment]
@NullVoxPopuli chaining IMO allows doing that in moar modular fashion tho' I agree that it's possible using one method with chained operations as you said
